### PR TITLE
Update resolution retry

### LINKED ISSUE #962: Handle update presence retries in the WorkerQueue
Right now we have a simple retry implemented for `presence.update()` inside the `Ably` wrapper class. However, this should be done in a more sophisticated way in the `WorkerQueue` (similarly to the presence enter and send message retries implemented in the `Publisher`). 

This work should affect the `Subscriber.requestResolution()` API method and publisher's `resolveResolution` internal method.

### PR DESCRIPTION
**Pull request by davyskiba**
Update resolution retries implemented on the worker queue level. 
Old approach annotated in subscriber as `deprecated.` After removing it, the old retry code can be removed.
Added `updatingResolutions` map to the `Properties` object to handle race scenarios when another update is sent before the scheduled update retry, resulting in an overwrite.

Resolves #962 #980

### CODE REVIEW

File: common/src/main/java/com/ably/tracking/common/Ably.kt [hunk 1]
```diff
@@ -773,14 +778,21 @@ constructor(
...
+            logHandler?.w("$TAG Failed to update presence data for trackable $trackableId", exception)
+            Result.failure(exception)
         }
     }
 
-    override suspend fun updatePresenceData(trackableId: String, presenceData: PresenceData): Result<Unit> {
+    override suspend fun updatePresenceDataWithRetry(trackableId: String, presenceData: PresenceData): Result<Unit> {
         val trackableChannel = getChannelIfExists(trackableId) ?: return Result.success(Unit)
         return try {
             // This simple retry mechanism should be removed while fixing https://github.com/ably/ably-asset-tracking-android/issues/962
```
Comment (AndyTWF) [inline common/src/main/java/com/ably/tracking/common/Ably.kt:798]: Can we remove this comment and/or retry mechanism now?

File: publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/workers/UpdateResolutionWorker.kt [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    override fun doWork(
+        properties: PublisherProperties,
+        doAsyncWork: (suspend () -> Unit) -> Unit,
+        postWork: (WorkerSpecification) -> Unit
+    ): PublisherProperties {
+        if (!properties.containsUpdatingResolution(trackableId, resolution)) {
+            properties.addUpdatingResolution(trackableId, resolution)
+        }
+        if (properties.isLastUpdatingResolution(trackableId, resolution)) {
+            doAsyncWork {
```
Comment (AndyTWF) [inline publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/workers/UpdateResolutionWorker.kt:26]: Should we move this `doAsyncWork` block into another method for readability - there's quite a lot of nesting going on here?

File: publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/workers/UpdateResolutionWorker.kt [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+                val result = waitAndUpdatePresence(properties.presenceData)
+                if (result.isFailure && !result.isFatalAblyFailure()) {
+                    postWork(WorkerSpecification.UpdateResolution(trackableId, resolution))
+                } else {
+                    postWork(WorkerSpecification.UpdateResolutionSuccess(trackableId, resolution))
+                }
+            }
+        } else {
+            properties.removeUpdatingResolution(trackableId, resolution)
```
Comment (AndyTWF) [inline publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/workers/UpdateResolutionWorker.kt:35]: Should we move this `else` above the rest and turn the `if/else` into a guard-style clause, save us a bit of nesting?

File: subscribing-sdk/src/main/java/com/ably/tracking/subscriber/workerqueue/workers/ChangeResolutionWorker.kt [hunk 4]
```diff
@@ -2,29 +2,47 @@ package com.ably.tracking.subscriber.workerqueue.workers
...
         properties: SubscriberProperties,
         doAsyncWork: (suspend () -> Unit) -> Unit,
         postWork: (WorkerSpecification) -> Unit
     ): SubscriberProperties {
         properties.presenceData = properties.presenceData.copy(resolution = resolution)
-
-        doAsyncWork {
-            val result = ably.updatePresenceData(trackableId, properties.presenceData)
-            callbackFunction(result)
+        if (!properties.containsUpdatingResolution(trackableId, resolution)) {
```
Comment (AndyTWF) [inline subscribing-sdk/src/main/java/com/ably/tracking/subscriber/workerqueue/workers/ChangeResolutionWorker.kt:22]: Same questions as the publisher re nesting and switching `else` for "guard" statements

File: common/src/main/java/com/ably/tracking/common/Ably.kt (hunk 1 above)
Comment (davyskiba) [reply common/src/main/java/com/ably/tracking/common/Ably.kt:798]: Unfortunately this is still needed for the deprecated subscriber methods, but the comment should be updated 🤔

Comment (davyskiba) [reply common/src/main/java/com/ably/tracking/common/Ably.kt]: Updated 8c948fd7095ef34d431a49af16a30dace0e0cf47

File: publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/workers/UpdateResolutionWorker.kt (hunk 2 above)
Comment (davyskiba) [reply publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/workers/UpdateResolutionWorker.kt]: Extract bee97931f70b43ff8bd6b09c281c55abcb0d07d4

File: subscribing-sdk/src/main/java/com/ably/tracking/subscriber/workerqueue/workers/ChangeResolutionWorker.kt (hunk 4 above)
Comment (davyskiba) [reply subscribing-sdk/src/main/java/com/ably/tracking/subscriber/workerqueue/workers/ChangeResolutionWorker.kt:22]: Extract bee97931f70b43ff8bd6b09c281c55abcb0d07d4

File: publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/workers/UpdateResolutionWorker.kt (hunk 3 above)
Comment (davyskiba) [reply publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/workers/UpdateResolutionWorker.kt:27]: 55acd50a3618e598225ed538cd50a7505ae71d35

**Review by AndyTWF — APPROVED** [review]

**Pull request merged**